File: algorithm/4월2주차/징검다리_최철원.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int solution(int distance, vector<int> rocks, int n) {
    int answer = 0;
    vector<int> v;
    sort(rocks.begin(),rocks.end());
    int cur=0;
    for(int i=0;i<rocks.size();i++){
        int loc = rocks[i];
        v.push_back(loc-cur);
        cur = loc;
    }
    v.push_back(distance-cur);
    int len = v.size();

    int l=0,r=1000000000,mid;
    while(l<=r){
        mid = l + (r-l)/2;
        int idx=0,cnt=0;
        while(idx<len){
            int val = v[idx];
            if(val<mid){        //합쳐야 하는 경우
                while(1){
                    if(cnt==n){     //더 이상 합치지 못하는 경우
                        cnt++;
                        break;
                    }
                    if(idx!=len-1){     //뒤로 합칠 수 있는 경우
                        int next = v[++idx];
                        val+=next;
                        cnt++;
                        if(val>=mid)    break;                        
                        else{       //뒤로 더 합쳐야 하는 경우
                            //계속 진행
                        }
                    }
                    else{       //앞으로 합쳐야 하는 경우(앞은 이미 기준 넘어섬)
                        cnt++;
                        break;
                    }
                }
            }
            else{       //기준 충족하는 경우
                
            }
            idx++;
        }
        if(cnt<=n){     
            l = mid+1;
            answer = mid;
        }
        else{       //합칠 갯수가 부족한 경우
            r = mid-1;
        }

    }
    return answer;
}

File: algorithm/4월2주차/징검다리_최철원.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int distance, int[] rocks, int n) {
        //초기화
        int answer = 0;
        int l = 0,r=1000000001,mid;
        List<Integer> li = new ArrayList<>();
        
        Arrays.sort(rocks);
        li.add(rocks[0]);
        for(int i=1;i<rocks.length;i++)
            li.add(rocks[i]-rocks[i-1]);
        li.add(distance - rocks[rocks.length-1]);
        int len = li.size();
        
        //바위간의 거리가 mid 이상의 거리가 아니라면 바위 제거
        while(l<=r){
            mid = l + (r-l)/2;
            int cnt = 0;
            int idx = 0;
            while(idx<len && cnt<=n){
                int a = li.get(idx++);
                if(a>=mid)  continue;
                else{
                    int sum = a;
                    while(idx<len){
                        sum+=li.get(idx++);
                        cnt++;
                        if(sum>=mid)    break;
                    }
                    if(sum<mid) cnt++;      //우측으로 끝까지 합해도 mid에 못미치는 경우, 앞이랑 합한다
                }
            }
            if(cnt>n)   r = mid-1;
            else if(cnt<=n){
                l= mid+1;
                answer = mid;         //해당 줄은 cnt==n일때만 해야 하는게 아닌가.. 문제 오류 아닌가?
            }
        }
        return answer;
    }
}
